### PR TITLE
Remove fake SSE support

### PR DESCRIPTION
**Pull request by tlively**
This is an immediate, useful step toward removing SIMD.js support entirely. SIMD.js is not supported on any platforms, and polyfilling it is often slower than using scalar code to begin with.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# For travis
```
Comment (dschuff) [inline Dockerfile]: Is this moved from somewhere else, or why is it in this CL?

**Comment by kripken** [comment]
cc @juj

**Comment by kripken** [comment]
Why remove the tests and benchmarks? We could disable the tests for now, but we do want to get them working eventually I think?

**Comment by juj** [comment]
No, don't remove this! This SSE support is not fake, and it is useful to be able to test compiled code against SIMD.js.

When there is proper support against Wasm SIMD, and the existing SSE compilation tests work against Wasm SIMD, then we can be happy to drop the SIMD.js path. Removing the path now just invites eager bitrot for SSE compilation support when it is not tested.

**Comment by tlively** [comment]
I'm not sure it is a good idea for the platform to pretend to be x86 when it's not. Users expect SSE intrinsics to have certain performance properties that may not hold when they are actually implemented with WebAssembly's SIMD. It would be bad to have unexpected performance cliffs in SIMD, which is a performance feature. If we do eventually provide SSE intrinsics in terms of wasm SIMD, I think it would be good to force users to specifically opt in to getting them by changing the flag or header names, as suggested by @dschuff. That way no one will accidentally pick up SSE intrinsics without understanding the performance implications.

@juj, even if we do eventually provide SSE instrinsics, the compilation path is completely different for asm.js and wasm, and the headers would have to be ported, so I'm not worried about bitrot. Since no platforms actually support SIMD.js, I'm not sure keeping support for it around is useful, even if wasm SIMD isn't quite ready yet.

@kripken I only removed tests that depended on the mocked SSE support. These tests will have to be ported to use the new wasm intrinsics anyway, so keeping them around in their present form is not very useful. They can be resurrected and ported once the wasm-native intrinsics have landed.

**Comment by tlively** [comment]
My personal motivation for this change is that I'm trying to port a library to use the wasm simd builtins as a POC and the fact that emscripten is pretending to support SSE is fouling up the build unnecessarily.

**Comment by dschuff** [comment]
The SSE support is fake, in fact it's is doubly-fake. It's fake because it's not actually SSE underneath, and it's doubly-fake because currently it generates SIMD.js and SIMD.js is not A Thing anymore; it's just some JS polyfill code. There is no value whatsoever in testing the correctness (let alone performance) of any SIMD.js codegen path. 

There could be some value in having a fake SSE implementation that has wasm SIMD underneath. This is something we've discussed and Thomas or Rich may well end up doing that if we think it's valuable (I'm not sure I really buy it but I could be convinced). In that case we could just bring back the relevant tests. But there's no reason to leave dead code checked into the repository until that happens.

**Comment by tlively** [comment]
@juj, does this change make more sense in light of our reasoning?

@kripken do you think removing the tests and benchmarks for now makes sense?

**Comment by kripken** [comment]
I'm fine with removing the tests and benchmarks since as you say they can be resurrected later. But I think the bigger question is how important they are - in the past they were very useful, and a lot of work went into them. So I think that's worth discussing here with @juj.

**Comment by juj** [comment]
> I'm not sure it is a good idea for the platform to pretend to be x86 when it's not. Users expect SSE intrinsics to have certain performance properties that may not hold when they are actually implemented with WebAssembly's SIMD.

> The SSE support is fake, in fact it's is doubly-fake. It's fake because it's not actually SSE underneath

> My personal motivation for this change is that I'm trying to port a library to use the wasm simd builtins as a POC and the fact that emscripten is pretending to support SSE is fouling up the build unnecessarily.

Developers need to be able to compile SSE intrinsics code for browsers running on x86 platforms, and NEON intrinsics code for browsers running on ARM platforms, in a way where browser cross-targets API usage for SSE compiled->NEON executed and NEON compiled->SSE executed wasm, or developer dual-hosts `page_sse.wasm` and `page_neon.wasm` code if automatic browser retargeting is not good enough. If anything else is provided, it will be a failure.

This is something that was battled ad nauseaum with SIMD.js, and people involved at the time just did not want to believe it, because they refused to believe that the right thing sometimes does not have the luxury of looking pretty.

Attempt to build up an API that is a set intersection of SSE and NEON consisting of "only convenient instructions" (instead of the set union), with no automated path to compile SSE/NEON code over, will become a thing where only one or two one or two poster projects will benefit where their SIMD code is hand-written from scratch to prove an imaginary point giving an impression that web now would have SIMD. In practice the world will ignore it as too much effort, unpredictable to develop against, and too weak benefit.

Anyhow, I was tired playing a broken record about this already two years ago, so I do not wish to interact with this aspect any more, and don't have too many extra cycles for this either. If you want to delete the SIMD&SSE paths and take the stance that Wasm SIMD support should not include a SSE target, feel free to go ahead with this PR if it can speed you up. I immensely respect you all, and the effort, but I just wish to turn my back on this disaster completely.

+1 for this to land, in whatever form it wants to take.

**Comment by dschuff** [comment]
A couple of responses to this.

Despite the fact that my previous response (reading it again) probably came off a bit harsher than I intended (and I am sorry about that):
I am not opposed to having SSE (and NEON) emulation (i.e. implementing SSE/NEON intrinsics on top of wasm) in the toolchain. I do think that it should be opt-in (i.e. the developer intentionally turns it on, knowing it will be emulated) because of the potentially interesting performance differences.
When that happens, the implementation and/or the tests could be derived from this code; we can do whatever makes the most practical sense there.

Regarding your broader point, it's looking like more of the folks involved with wasm and SIMD may be coming around to share your view, and that future SIMD extensions may have more CPU-specific flavors. It's certainly true that we won't be able to expose the full performance of today's CPUs without doing that. But there's also a desire to finalize and ship what we've gotten so far first.

**Comment by penzn** [comment]
It may be possible to a few optional target features, corresponding to classes of hardware intrinsics to be emulated in WASM to Clang and do the needful under those. We are currently evaluating this.

**Comment by abrown** [comment]
A week or two ago I was trying to use emscripten to compile an example with `xmmintrin.h` intrinsics to WASM and @sunfishcode mentioned that this feature had been removed but could possibly be revived. @tlively, could we revive this? I certainly would find it useful.

**Comment by tlively** [comment]
This comes up as a request every once in a while. Here are the two usual solutions people have ended up with.

 1. **You care enough about speed to do work:** Then you should port the performance-critical portions of your code to WebAssembly-native SIMD using the wasm_simd128.h header for maximum speed.
 1. **You don't care about speed, but your code requires xmmintrin.h**: You can get emulated SIMD using [SIMDe](https://github.com/nemequ/simde). This is good because it's unmistakably an emulation.

If your use case isn't covered by one of those, I'd be interested in hearing more about it.

It's possible that implementing the SSE intrinsics in terms of WebAssembly would give good results after all, but we would want to validate that thoroughly before providing that support by default, and that hasn't been as high a priority as driving the SIMD spec proposal forward.

**Pull request merged**